addons/debugger/module.js

```diff
@@ -0,0 +1,378 @@
+  for (var i = startIndex; i < vm.runtime.threads.length; i++) {
+    const newThread = vm.runtime.threads[i];
+
+    if (newThread.status === STATUS_YIELD_TICK) {
+      newThread.status = STATUS_RUNNING;
+    }
+
+    if (newThread.status === STATUS_RUNNING || newThread.status === STATUS_YIELD) {
+      return newThread;
+    }
+  }
+  return null;
+};
+
+export const singleStep = () => {
+  const pauseState = pausedThreadState.get(steppingThread);
+  if (steppingThread) {
+    // Make it look like no time has passed
+    const stackFrame = steppingThread.peekStackFrame();
+    if (stackFrame && stackFrame.executionContext && stackFrame.executionContext.timer) {
+      stackFrame.executionContext.timer.startTime += vm.runtime.currentMSecs - pauseState.time;
+    }
+    pauseState.time = vm.runtime.currentMSecs;
+
+    // Execute the block
+    const continueExecuting = singleStepThread(steppingThread);
+
+    if (!continueExecuting) {
+      // Try to move onto the next thread
+      steppingThread = findNewSteppingThread(vm.runtime.threads.indexOf(steppingThread) + 1);
+    }
+  }
+
+  // If we don't have a thread, than we are between VM steps and should search for a new thread
+  if (!steppingThread) {
+    steppingThread = findNewSteppingThread(0);
+
+    // End of VM step, emulate one frame of time passing.
+    vm.runtime.ioDevices.clock._pausedTime += vm.runtime.currentStepTime;
+    // Skip all sounds forward by vm.runtime.currentStepTime miliseconds so it's as
+    //  if they where playing for one frame.
+    const audioContext = vm.runtime.audioEngine.audioContext;
+    for (const target of vm.runtime.targets) {
+      for (const soundId in target.sprite.soundBank.soundPlayers) {
+        const soundPlayer = target.sprite.soundBank.soundPlayers[soundId];
+        if (soundPlayer.outputNode) {
+          soundPlayer.outputNode.stop(audioContext.currentTime);
+          soundPlayer._createSource();
+          soundPlayer.outputNode.start(
+            audioContext.currentTime,
+            audioContext.currentTime - soundPlayer.startingUntil + vm.runtime.currentStepTime / 1000
+          );
+          soundPlayer.startingUntil -= vm.runtime.currentStepTime / 1000;
+        }
+      }
+    }
+    // Move all threads forward one frame in time. For blocks like `wait () seconds`
+    for (const thread of vm.runtime.threads) {
+      if (pausedThreadState.has(thread)) {
+        pausedThreadState.get(thread).time += vm.runtime.currentStepTime;
+      }
+    }
+
+    // Try to run edge activated hats
+    pauseNewThreads = true;
+
+    const hats = vm.runtime._hats;
+    for (const hatType in hats) {
+      if (!hats.hasOwnProperty(hatType)) continue;
+      const hat = hats[hatType];
+      if (hat.edgeActivated) {
+        vm.runtime.startHats(hatType);
+      }
+    }
+
+    pauseNewThreads = false;
+  }
+
+  eventTarget.dispatchEvent(new CustomEvent("step"));
+};
+
+export const setup = (addon) => {
```

Comment:
It would be cleaner to make the vm be the parameter, instead of the whole addon object. Looks like only the vm is used on this function.